core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```diff
@@ -639,7 +641,11 @@ class GroupCoordinator(val brokerId: Int,
               responseCallback(Errors.UNKNOWN_MEMBER_ID)
 
             case CompletingRebalance =>
-                responseCallback(Errors.REBALANCE_IN_PROGRESS)
+              // consumers may start sending heartbeat after join-group response, in which case
+              // we should treat them as normal hb request and reset the timer
+              val member = group.get(memberId)
```

Comment:
This is the only logical change as 3) in the description. All others are logging changes.

Reply:
Wait, so before this the coordinator wouldn't complete the current heartbeat? Doesn't that mean that heartbeating is pointless until the rebalance completes? Obviously that doesn't line up with my observations since members were clearly getting kicked from the group before the rebalance had completed, so I must be missing something here

Reply:
It would return the error code before: that is because it does not expect clients to send heartbeat before sending sync-group requests. Now it is not the case any more.